### PR TITLE
Add HPET and I/O APIC setup code

### PR DESCRIPTION
**Pull request by Rwantare**
WIP because the problem isn't fixed yet

### CODE REVIEW

**Comment by rminnich** [comment]
yeah but this could be what we want anyway? It seems fine.

File: src/mainboard/amd/romecrb/src/acpi.rs [hunk 1]
```diff
@@ -11,44 +11,51 @@ use util::round_up_4k;
...
+    // let madt_offset = dsdt_offset + 69887;
     let madt_offset = dsdt_offset + DSDT_DSDTTBL_HEADER.len();
     let madt_local_apic_offset = madt_offset + size_of::<AcpiTableMadt>();
-    let io_apic_offset = madt_local_apic_offset + cores as usize * size_of::<AcpiMadtLocalApic>();
-    let local_x2apic_offset = io_apic_offset + size_of::<AcpiMadtIoApic>();
-    let local_lapicnmi_offset = local_x2apic_offset + cores as usize * size_of::<AcpiMadtLocalX2apic>();
-    let local_isor_offset = local_lapicnmi_offset + size_of::<AcpiMadtLocalX2ApicNMI>();
-    let mcfg_offset = local_isor_offset + 2 * size_of::<AcpiMadtInterruptOverride>();
-    let total_size = mcfg_offset + size_of::<AcpiTableMcfg>() - start;
+    let madt_local_x2apic_offset = madt_local_apic_offset + cores as usize * size_of::<AcpiMadtLocalApic>();
```
Comment (fotonick) [inline src/mainboard/amd/romecrb/src/acpi.rs:28]: Could a crate like [memoffset](https://docs.rs/memoffset/0.6.1/memoffset/macro.offset_of.html) be useful here?

Comment (fotonick) [reply src/mainboard/amd/romecrb/src/acpi.rs:28]: Probably much better would be to generate an SVD file, then use [svd2rust](https://docs.rs/svd2rust/0.17.0/svd2rust/) to generate register descriptions for us, but this is probably a fair bit of tedious work, then a major refactor.

Comment (rminnich) [reply src/mainboard/amd/romecrb/src/acpi.rs:28]: I tihnk the day is coming when we rework the way we do this, and hopefully in a way we can take it back to xhype.

File: src/mainboard/amd/romecrb/src/ioapic.rs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        ptr::write_volatile((ioapic_base + 10) as *mut u32, value);
+    }
+}
+
+fn ioapic_interrupt_count(ioapic_base: u32) -> u32 {
+    let mut ioapic_interrupts = (ioapic_read(ioapic_base, 1) >> 16) & 0xFF;
+    if ioapic_interrupts == 0xFF {
+        ioapic_interrupts = 23;
+    }
+    ioapic_interrupts += 1;
```
Comment (rminnich) [inline src/mainboard/amd/romecrb/src/ioapic.rs:22]: is this one of those times you could just say 
ioapic_interrupts + 1
and skip the ; and the return?

Comment (rminnich) [reply src/mainboard/amd/romecrb/src/ioapic.rs:22]: oh and further (I'm learning)
if io_apic_interrupts == 0xff {
23
} else {
ioapic_interrupts + 1
}
?
or return 23; 

just wondering. Then it need not me mut, not that it matters?

File: src/mainboard/amd/romecrb/src/ioapic.rs [hunk 3]
```diff
@@ -0,0 +1,61 @@
+use core::ptr;
+
+fn ioapic_read(ioapic_base: u32, register: u32) -> u32 {
```
Comment (rminnich) [inline src/mainboard/amd/romecrb/src/ioapic.rs:3]: TODO: separate the basic generic ioapic_read and _write (can't it be a driver?) from the specific details for this board

File: src/mainboard/amd/romecrb/src/mainboard.rs [hunk 4]
```diff
@@ -30,17 +30,17 @@ const FCH_LEGACY_3F8_SH: u16 = 1 << 3;
 // 2. It would be JUST ONE MORE.
 // SMN
 
-fn snmr(a: u32) -> u32 {
+fn smn_read(a: u32) -> u32 {
```
Comment (rminnich) [inline src/mainboard/amd/romecrb/src/mainboard.rs:33]: I wonder why I did not make smn a driver/

**Review by rminnich — APPROVED** [review]
see comments, yours to change or tidy up as you feel best.

**Review by rminnich — APPROVED** [review]

**Review by rminnich — APPROVED** [review]

File: src/mainboard/amd/romecrb/src/acpi.rs [hunk 5]
```diff
@@ -11,44 +11,51 @@ use util::round_up_4k;
...
-    let xsdt = AcpiTableHeader { signature: SIG_XSDT, length: xsdt_total_length as u32, ..AcpiTableHeader::new() };
+    let xsdt = AcpiTableHeader { signature: SIG_XSDT, length: xsdt_total_length as u32, revision: 1, ..AcpiTableHeader::new() };
     write(w, xsdt, xsdt_offset, 0);
     // xsdt entries
     let mut xsdt_entries: [u64; NUM_XSDT_ENTRIES] = [0; NUM_XSDT_ENTRIES];
     xsdt_entries[0] = fadt_offset as u64;
     xsdt_entries[1] = madt_offset as u64;
-    // xsdt_entries[2] = mcfg_offset as u64;
+    xsdt_entries[2] = mcfg_offset as u64;
+    xsdt_entries[3] = hpet_offset as u64;
```
Comment (rjoleary) [inline src/mainboard/amd/romecrb/src/acpi.rs:57]: Could these 4 lines be in an array literal, so xsdt_entries doesn't need to be mut?

File: src/mainboard/amd/romecrb/src/acpi.rs [hunk 6]
```diff
@@ -92,50 +103,64 @@ pub fn setup_acpi_tables(w: &mut impl core::fmt::Write, start: usize, cores: u32
     debug_assert_eq!(acpi_tb_checksum(fadt_offset, fadt_offset + size_of::<AcpiTableFadt>()), 0);
 
     // facs - Firmware ACPI Control Structure
-    let facs = AcpiTableFacs { signature: SIG_FACS, length: 64, flags: 0b0011, version: 2, ..Default::default() };
+    let facs = AcpiTableFacs { signature: SIG_FACS, length: size_of::<AcpiTableFacs>() as u32, flags: 0, version: 2, ..Default::default() };
     write(w, facs, facs_offset, 0);
 
-    // dsdt - Differentiated System Description Table
+    // dsdt - Differentiated System Description Table ******!!!!!!!!!!!
```
Comment (rjoleary) [inline src/mainboard/amd/romecrb/src/acpi.rs:109]: What is the `******!!!!!!!!!!!` for?

File: src/mainboard/amd/romecrb/src/acpi.rs [hunk 7]
```diff
@@ -92,50 +103,64 @@ pub fn setup_acpi_tables(w: &mut impl core::fmt::Write, start: usize, cores: u32
...
     write(w, DSDT_DSDTTBL_HEADER, dsdt_offset, 0);
 
+    // let DSDT = include_bytes!("../acpi/dsdt.dat");
+    // write(w, DSDT, dsdt_offset, 0);
```
Comment (rjoleary) [inline src/mainboard/amd/romecrb/src/acpi.rs:113]: Remove this

File: src/mainboard/amd/romecrb/src/acpi.rs (hunk 6 above)
Comment (Rwantare) [reply src/mainboard/amd/romecrb/src/acpi.rs:109]: A bookmark I forgot to remove

**Review by rjoleary — APPROVED** [review]

**Pull request merged**